### PR TITLE
snippets/ev3: add PS4 snippet

### PR DESCRIPTION
**Pull request by codeadamca**
Added an EV3 snippet to use a PS4 controller with a few motors in a tank like style. The snippet also includes print commands associated to most of the common PS4 buttons. For a full list of buttons, touchpad, and motion checkout: https://github.com/codeadamca/python-connect-ps4

### CODE REVIEW

File: snippets/ev3/ps4/main.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env pybricks-micropython
+from pybricks.ev3devices import (Motor)
+from pybricks.parameters import Port, Stop, Direction, Button, Color
+from pybricks.tools import wait, StopWatch, DataLog
```
Comment (laurensvalk) [inline snippets/ev3/ps4/main.py:4]: You can omit the () around Motor. Can you also remove any unused imports, like DataLog?

File: snippets/ev3/ps4/main.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+# For a full map of all PS4 buttons, trackpad, and motion check  out:
+# https://github.com/codeadamca/python-connect-ps4
+
+# Create your objects here.
+left_motor = Motor(Port.B)
+right_motor = Motor(Port.C)
+left_speed = 0
+right_speed = 0
+
+# Locat the event file you want to react to, on my setup the PS4 controller button events
```
Comment (laurensvalk) [inline snippets/ev3/ps4/main.py:18]: Locate

File: snippets/ev3/ps4/main.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+FORMAT = 'llHHi'
+EVENT_SIZE = struct.calcsize(FORMAT)
+event = in_file.read(EVENT_SIZE)
+
+# A helper function for converting stick values (0 to 255) to more usable numbers (-100 to 100)
+def scale(val, src, dst):
+    return (float(val - src[0]) / (src[1] - src[0])) * (dst[1] - dst[0]) + dst[0]
+
+# Create a loop to react to events
+# This loop react to all main PS4 button and sstick events. I ave left out buttons like
```
Comment (laurensvalk) [inline snippets/ev3/ps4/main.py:34]: reacts
stick
have

File: snippets/ev3/ps4/main.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            print("The horizontal directional pad is released")
+
+    # Set motor speed
+    left_motor.dc(left_speed)
+    right_motor.dc(right_speed)
+
+    # Read the next event
+    event = in_file.read(EVENT_SIZE)
+
+in_file.close()
```
Comment (laurensvalk) [inline snippets/ev3/ps4/main.py:139]: Could you add an empty new line?

You can have VSCode do this automatically for you:

image

Comment (laurensvalk) [reply snippets/ev3/ps4/main.py:139]: And finally, can you change the commit message to something like this to be consistent with the existing log?

snippets/ev3: add PS4 snippet

File: snippets/ev3/ps4/main.py (hunk 1 above)
Comment (codeadamca) [reply snippets/ev3/ps4/main.py:4]: Done!

**Comment by codeadamca** [comment]
Changes have been addressed!

**Comment by laurensvalk** [comment]
Thanks, but I'm not sure the comment fixes got through. This is the only change I see.

image

Also please run `flake8`, which will give some tips for further cleanup:

```
./snippets/ev3/ps4/main.py:8:89: E501 line too long (101 > 88 characters)
./snippets/ev3/ps4/main.py:18:89: E501 line too long (89 > 88 characters)
./snippets/ev3/ps4/main.py:29:89: E501 line too long (95 > 88 characters)
./snippets/ev3/ps4/main.py:30:1: E302 expected 2 blank lines, found 1
./snippets/ev3/ps4/main.py:35:81: W291 trailing whitespace
./snippets/ev3/ps4/main.py:37:1: E305 expected 2 blank lines after class or function definition, found 1
./snippets/ev3/ps4/main.py:92:1: W293 blank line contains whitespace
./snippets/ev3/ps4/main.py:93:23: E261 at least two spaces before inline comment
./snippets/ev3/ps4/main.py:95:89: E501 line too long (102 > 88 characters)
./snippets/ev3/ps4/main.py:95:103: W291 trailing whitespace
./snippets/ev3/ps4/main.py:97:1: W293 blank line contains whitespace
./snippets/ev3/ps4/main.py:98:43: W291 trailing whitespace
./snippets/ev3/ps4/main.py:100:51: E231 missing whitespace after ','
./snippets/ev3/ps4/main.py:101:41: E231 missing whitespace after ','
./snippets/ev3/ps4/main.py:102:1: W293 blank line contains whitespace
./snippets/ev3/ps4/main.py:103:45: W291 trailing whitespace
./snippets/ev3/ps4/main.py:105:53: E231 missing whitespace after ','
./snippets/ev3/ps4/main.py:109:52: E231 missing whitespace after ','
./snippets/ev3/ps4/main.py:110:42: E231 missing whitespace after ','
./snippets/ev3/ps4/main.py:114:54: E231 missing whitespace after ','
./snippets/ev3/ps4/main.py:115:1: W293 blank line contains whitespace
./snippets/ev3/ps4/main.py:117:9: E303 too many blank lines (2)
./snippets/ev3/ps4/main.py:139:16: W292 no newline at end of file
```

**Comment by codeadamca** [comment]
I think it's all good now.

**Comment by laurensvalk** [comment]
Thank you for the quick updates. I squashed and merged it. I am somewhat tempted to get a PS4 controller now :smile:

**Comment by codeadamca** [comment]
Yea it's fun! I've connected them to LEGO using a BrickPi, but never got it working using the EV3 Brick until now.

That's my first time contributing to another repository, thanks for your patience!!

Adam
________________________________

**Comment by laurensvalk** [comment]
Cool - I think this my first time squashing and force pushing over someone's PR, so that makes two of us. I think I managed to not accidentally delete it all.

I wonder if the upcoming PS5 controller could technically work, and whether it has BLE so it might work with the newer LEGO hubs too (most of which don't have Bluetooth Classic like EV3).

**Comment by codeadamca** [comment]
The EV3 is the only hub I have. I'd love to try once I can get my hands on a PS5 controller and some other LEGO hubs.

**Pull request merged**